File: Python/linked_lists/delete_middle_node.py
# Import the LinkedList class from the linked_list module
from linked_list import LinkedList

# Define a function that deletes a middle node from a linked list
def delete_middle_node(node):
    # Replace the value of the input node with the value of the next node
    node.value = node.next.value
    # Update the next pointer of the input node to skip the next node and point to the one after it
    node.next = node.next.next

# If this script is run directly (not imported as a module), do the following:
if __name__ == "__main__":
    # Create a new linked list instance
    ll = LinkedList()
    # Add multiple nodes to the linked list with values [1, 2, 3, 4]
    ll.add_multiple([1, 2, 3, 4])
    # Add a new node to the linked list with value 5, and store a reference to it in middle_node
    middle_node = ll.add(5)
    # Add multiple nodes to the linked list with values [7, 8, 9]
    ll.add_multiple([7, 8, 9])

    # Print the linked list before the middle node is deleted
    print(ll)
    # Call the delete_middle_node function on the middle node
    delete_middle_node(middle_node)
    # Print the linked list after the middle node is deleted
    print(ll)

File: C++/linked_lists/delete_middle_node.cpp
#include <iostream>
// This code defines a singly linked list node structure and provides two helper functions, `printList()` and `deleteNode()`.

// Definition of a singly linked list node structure
struct Node {
  char data; // The data value stored in the node
  Node * next; // Pointer to the next node in the list

  // Constructor for Node structure
  Node( char c ) : data{ c }, next{ nullptr } { }
};

/**
 * Prints the data values for each node in the linked list
 * @param head Pointer to the head of the linked list
 */
void printList( Node * head ) {
  while( head ) { // loop while the head pointer is not null
    std::cout << head->data << "-->"; // output the data value of the current node, followed by an arrow
    head = head->next; // move to the next node
  }
  std::cout << "nullptr" << std::endl; // output "nullptr" to indicate the end of the list
}

/**
 * Deletes a node from the linked list
 * @param node Pointer to the node to be deleted
 */
void deleteNode( Node * node ) {
  if ( node == nullptr || node->next == nullptr ) { // check if the given node is null or the last node in the list
    return; // if so, return without doing anything
  }
  Node * nextNode = node->next; // store a pointer to the next node in the list
  node->data = nextNode->data; // copy the data value of the next node into the current node
  node->next = nextNode->next; // set the next pointer of the current node to skip the next node
  delete nextNode; // delete the next node from memory
}


/**
 * Main function to demonstrate usage of Node structure and helper functions
 */
int main() {
  Node * head = new Node('a'); // create a new node 'a' and set it as the head of the list
  head->next = new Node('b'); // create a new node 'b' and set it as the next node in the list
  head->next->next = new Node('c'); // create a new node 'c' and set it as the next node in the list
  head->next->next->next = new Node('d'); // create a new node 'd' and set it as the next node in the list
  head->next->next->next->next = new Node('e'); // create a new node 'e' and set it as the next node in the list

  std::cout << "List before deletion:\n";
  printList(head); // print the original list using the printList() function

  std::cout << "Deleting node containing data as 'c'\n";
  deleteNode(head->next->next); // delete the node containing data value 'c' using the deleteNode() function

  std::cout << "List after deletion:\n";
  printList(head); // print the resulting list after the deletion using printList() function

  return 0; // exit the program
}
